File: CodingTest/Coding_Test/SW Expert Academy(SWEA)/D4/1868. 파핑파핑 지뢰찾기.java
import java.util.*;
import java.lang.*;
import java.io.*;

class Tuple{
    private int first;
    private int second;

    public Tuple(int first,int second){
        this.first=first;
        this.second=second;
    }

    public int getFirst(){
        return this.first;
    }

    public int getSecond(){
        return this.second;
    }
}

// The main method must be in a class named "Main".
class Main {
    public static void main(String[] args) {
        Scanner sc=new Scanner(System.in);
        int tc=sc.nextInt();
        int[] dx={0,0,-1,1,-1,-1,1,1};
        int[] dy={-1,1,0,0,1,-1,1,-1};

        for(int i=1;i<=tc;i++){
            int answer=0;
            int n=sc.nextInt();
            String[][] board=new String[n][n];
            int[][] bomb_checker=new int[n][n];
            boolean[][] visited=new boolean[n][n];
            for(int j=0;j<n;j++){
                String s=sc.next();
                for(int k=0;k<n;k++){
                    board[j][k]=Character.toString(s.charAt(k));
                }    
            }
            for(int j=0;j<n;j++){
                for(int k=0;k<n;k++){
                    if(board[j][k].equals("*")){
                        bomb_checker[j][k]=-1;
                    }else{
                        int cnt=0;
                        for(int m=0;m<8;m++){
                            int nx=j+dx[m];
                            int ny=k+dy[m];
                            if(nx<0 || ny<0 || nx>=n || ny>=n){
                                continue;
                            }
                            if(board[nx][ny].equals("*")){
                                cnt++;
                            }
                        }
                        bomb_checker[j][k]=cnt;
                    }
                }
            }

            for(int j=0;j<n;j++){
                for(int k=0;k<n;k++){
                    if(bomb_checker[j][k]==0 && !visited[j][k]){
                        answer++;
                        visited[j][k]=true;
                        Queue<Tuple> q=new LinkedList<>();
                        q.offer(new Tuple(j,k));
                        while(!q.isEmpty()){
                            Tuple tuple=q.poll();
                            int x=tuple.getFirst();
                            int y=tuple.getSecond();
                            if(bomb_checker[x][y]>0){
                                continue;
                            }
                            for(int m=0;m<8;m++){
                                int nx=x+dx[m];
                                int ny=y+dy[m];
                                if(nx<0||ny<0||nx>=n||ny>=n){
                                    continue;
                                }
                                if(visited[nx][ny]){
                                    continue;
                                }
                                if(bomb_checker[nx][ny]==-1){
                                    continue;
                                }
                                if(m>=4){
                                    if(!visited[nx][y] || !visited[x][ny]){
                                        continue;
                                    }
                                }
                                visited[nx][ny]=true;
                                q.offer(new Tuple(nx,ny));
                            }
                        }
                    }
                }
            }
            for(int j=0;j<n;j++){
                for(int k=0;k<n;k++){
                    if(!visited[j][k] && bomb_checker[j][k]>0){
                        answer++;
                    }
                }
            }
            System.out.printf("#%d %d\n",i,answer);
        }

    }
}

File: CodingTest/Coding_Test/SW Expert Academy(SWEA)/D4/1868. 파핑파핑 지뢰찾기.py
import copy
from collections import deque

for tc in range(int(input())):
    n=int(input())
    board=[]
    dx=[0,0,-1,1,-1,-1,1,1]
    dy=[-1,1,0,0,-1,1,-1,1]
    for i in range(n):
        board.append(list(input()))

    bomb_checker=[[0]*n for _ in range(n)]
    for i in range(n):
        for j in range(n):
            if board[i][j]=='*':
                bomb_checker[i][j]=-1
                continue
            cnt=0
            for k in range(8):
                nx=i+dx[k]
                ny=j+dy[k]
                if nx<0 or ny<0 or nx>=n or ny>=n:
                    continue
                if board[nx][ny]=='*':
                    cnt+=1
            bomb_checker[i][j]=cnt
    answer=0
    visited=[[False]*n for _ in range(n)]
    for i in range(n):
        for j in range(n):
            if bomb_checker[i][j]==0 and not visited[i][j]:
                q=deque([])
                q.append((i,j))
                visited[i][j]=True
                answer+=1
                while q:
                    x,y=q.popleft()
                    if bomb_checker[x][y]>0:
                        continue
                    for k in range(8):
                        nx=x+dx[k]
                        ny=y+dy[k]
                        if nx<0 or ny<0 or nx>=n or ny>=n:
                            continue
                        if visited[nx][ny]:
                            continue
                        if bomb_checker[nx][ny]==-1:
                            continue
                        if k>=4:
                            if not visited[x][ny] or not visited[nx][y]:
                                continue
                        visited[nx][ny]=True
                        q.append((nx,ny))
    for i in range(n):
        for j in range(n):
            if bomb_checker[i][j]>0 and not visited[i][j]:
                answer+=1

    print("#"+str(tc+1)+" "+str(answer))

